packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,338 @@
+
+    elected = firstElected;
+  }
+
+  // TODO maybe "freezeEpochRewards" "prepareForNextEpoch"
+
+  /// start next epoch process.
+  /// it freezes the epochrewards at the time of execution,
+  /// and starts the distribution of the rewards.
+  function startNextEpochProcess() external nonReentrant {
+    require(systemAlreadyInitialized(), "Epoch system not initialized");
+    require(isTimeForNextEpoch(), "Epoch is not ready to start");
+    require(!isOnEpochProcess(), "Epoch process is already started");
+    epochProcessing.status = EpochProcessStatus.Started;
+
+    epochs[currentEpochNumber].rewardsBlock = block.number;
+
+    // calculate rewards
+    getEpochRewards().updateTargetVotingYield();
+
+    (
+      uint256 perValidatorReward,
+      uint256 totalRewardsVoter,
+      uint256 totalRewardsCommunity,
+      uint256 totalRewardsCarbonFund
+    ) = getEpochRewards().calculateTargetEpochRewards();
+
+    epochProcessing.perValidatorReward = perValidatorReward;
+    epochProcessing.totalRewardsVoter = totalRewardsVoter;
+    epochProcessing.totalRewardsCommunity = totalRewardsCommunity;
+    epochProcessing.totalRewardsCarbonFund = totalRewardsCarbonFund;
+
+    allocateValidatorsRewards();
+
+    emit EpochProcessingStarted(currentEpochNumber);
+  }
+
+  function finishNextEpochProcess(
+    address[] calldata groups,
+    address[] calldata lessers,
+    address[] calldata greaters
+  ) external nonReentrant {
+    require(isOnEpochProcess(), "Epoch process is not started");
+    // finalize epoch
+    // last block should be the block before and timestamp from previous block
+    epochs[currentEpochNumber].lastBlock = block.number - 1;
+    // start new epoch
+    currentEpochNumber++;
+    epochs[currentEpochNumber].firstBlock = block.number;
+    epochs[currentEpochNumber].startTimestamp = block.timestamp;
+
+    for (uint i = 0; i < elected.length; i++) {
+      address group = getValidators().getValidatorsGroup(elected[i]);
+      if (!processedGroups[group].processed) {
+        epochProcessing.toProcessGroups++;
+        uint256 groupScore = getScoreReader().getGroupScore(group);
+        // We need to precompute epoch rewards for each group since computation depends on total active votes for all groups.
+        uint256 epochRewards = getElection().getGroupEpochRewards(
+          group,
+          epochProcessing.totalRewardsVoter,
+          groupScore
+        );
+        processedGroups[group] = ProcessedGroup(true, epochRewards);
+      }
+    }
+
+    require(epochProcessing.toProcessGroups == groups.length, "number of groups does not match");
+
+    // since we are adding values it makes sense to start from the end
+    for (uint ii = groups.length; ii > 0; ii--) {
+      uint256 i = ii - 1;
+      ProcessedGroup storage processedGroup = processedGroups[groups[i]];
+      // checks that group is actually from elected group
+      require(processedGroup.processed, "group not processed");
```

Comment:
better to say "group not from current elected set"

Reply:
addressed by #11223 commit cbf5fdc59053c4a5162354dcb6058476e276706e